sdk/core/Azure.Core/src/Pipeline/Internal/RequestScopedHttpMessageProperties.cs

```diff
@@ -0,0 +1,66 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+
+namespace Azure.Core.Pipeline
+{
+    internal static class RequestScopedHttpMessageProperties
+    {
+        private static readonly AsyncLocal<HttpMessagePropertiesScope?> CurrentHttpMessagePropertiesScope = new AsyncLocal<HttpMessagePropertiesScope?>();
```

Comment:
Should we move this async local onto the HttpPipeline itself?

Reply:
I started it separately because I didn't want to bloat the HttpPipeline class too much. I'm assuming that if we move this to HttpPipeline then the surroundings also go there. Would you prefer that ?

Reply:
Yeah, I think these two are pretty tightly coupled so it would make sense to move this logic into HttpPipeline.